azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoaderImpl.java

```diff
@@ -0,0 +1,217 @@
+package azkaban.version;
+
+import azkaban.db.DatabaseOperator;
+import azkaban.db.SQLTransaction;
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.apache.log4j.Logger;
+
+/**
+ * A JDBC based implementation for loading VersionSet from the backed db.
+ * This class keeps the copy of entries in the version_set table locally
+ * as well to reduce the number of queries to the db.
+ *
+ * This is a singleton class to be instantiated using dependency injection.
+ */
+@Singleton
+public class JdbcVersionSetLoaderImpl implements VersionSetLoader {
+    private static final Logger logger = Logger.getLogger(JdbcVersionSetLoaderImpl.class);
+
+    private static final String SELECT_VSET_FROM_NUM = "SELECT vset.num, vset.md5, vset.json FROM version_set vset WHERE vset.num=?";
+    private static final String SELECT_VSET_FROM_MD5 = "SELECT vset.num, vset.md5, vset.json FROM version_set vset WHERE vset.md5=?";
+    private static final String SELECT_ALL_VSET = "SELECT vset.num, vset.md5, vset.json FROM version_set vset";
+    private static final String INSERT_VSET = "INSERT INTO version_set (md5, json) values (?,?)";
+    private static final String DELETE_VSET = "DELETE FROM version_set vset WHERE vset.md5=?";
```

Comment:
version_set has no concept of active/deprecated. 
It is only stored in the db to identify if the exact combination was used earlier as well and hence to determine the id allocated for that. 

Deletion functionality may never be utilized by added so that in future if the table requires clean up, then this can be leveraged.

Reply:
What if there is any image version which is faulty. There is version set already populated with that version. User is using version set number in flow parameter , how will we stop them from using faulty version of any image?

Reply:
Reusing the same version set is a functionality added only to provide the user with the same environment as of some previous execution if it is mentioned explicitly, and hence if the user wants to use the same version set, it should be allowed. If this is not explicitly provided, then anyway the versions will be calculated at runtime and the deprecated jobtype version won't be utilized.